forc-pkg/src/manifest.rs

```diff
@@ -68,6 +68,7 @@ impl Manifest {
     pub fn from_file(path: &Path) -> anyhow::Result<Self> {
         let manifest = std::fs::read_to_string(path)
             .map_err(|e| anyhow!("failed to read manifest at {:?}: {}", path, e))?;
+        warn_unused_key(&manifest)?;
         let manifest: Self =
             toml::from_str(&manifest).map_err(|e| anyhow!("failed to parse manifest: {}.", e))?;
```

Comment:
Here we end up deserializing the string to `Manifest` twice: once for the unused key check, and once afterwards for the `manifest` that we end up using. It looks like `serde_ignored` was designed in a manner that should let us avoid this.

E.g. we should be able to do something like:

```rust
        let manifest_str = std::fs::read_to_string(path)
            .map_err(|e| anyhow!("failed to read manifest at {:?}: {}", path, e))?;
        let toml_de = &mut toml::Deserializer::from_str(&manifest_str);
        let manifest: Self = serde_ignored::deserialize(toml_de, |path| {
            let warning = format!("  WARNING! unused manifest key: {}", path.to_string());
            println_yellow_err(&warning).unwrap();
        })?;
        // ...
```

Reply:
Wouldn't this conflict with your error mapping? And also would this print a warning for each unused path?

Reply:
> Wouldn't this conflict with your error mapping?

It shouldn't conflict with the error mapping, I just forgot to add it on the end of that snippet. Here's roughly what it would look like with the original error mapping included:
```rust
        let manifest_str = std::fs::read_to_string(path)
            .map_err(|e| anyhow!("failed to read manifest at {:?}: {}", path, e))?;
        let toml_de = &mut toml::Deserializer::from_str(&manifest_str);
        let manifest: Self = serde_ignored::deserialize(toml_de, |path| {
            let warning = format!("  WARNING! unused manifest key: {}", path.to_string());
            println_yellow_err(&warning).unwrap();
        }).map_err(|e| anyhow!("failed to parse manifest: {}.", e))?;
        // ...
```

> also would this print a warning for each unused path?

Yes, this just prints them directly rather than first collecting them into a set, and then iterating over the set and printing them afterwards.

Reply:
Would rather have them in the order in which they occur!